### PR TITLE
fix: Derive generic types

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves an issue alluded to in https://github.com/noir-lang/noir/pull/5667 where we still could not derive generic types

## Summary\*

This PR allows us to derive generic types by allowing a new "resolved generic" type in the UnresolvedGeneric enum.

The issue before was that when we splice `Type`s into quoted snippets we lower them as a special token kind which preserves the original `Type` so that we don't have to worry about re-resolving it if it is a struct type that is not imported into the caller's scope, etc. This also means though that any type variables in the original type are preserved. In the following snippet:

```rs
// Lets say this is `MyType<T>`
let my_type: Type = ...;
quote {
    impl<T> Foo for $my_type { ... }
}
```

The impl will introduce a fresh generic `T` into scope, but the `T` in `MyType<T>` will still be the old `T` with a different TypeVariableId. To fix this there were two options:
- Add another function to lower types in a way they'd need to be reparsed, and would thus pick up new type variables but be susceptible to "name not in scope" errors mentioned earlier.
- Somehow get the `T` in `impl<T>` to refer to the same `T` in `MyType<T>`.

I chose the later approach since it is much simpler from a user's perspective and leads to fewer footguns over forgetting to convert the type into a `Quoted` before splicing it into other code. This way, users just need to ensure the generics inserted as part of the impl generics come from the type which is what is typically done anyway. This means that the original snippet above will actually still fail since it still introduces a new `T`. To fix it, we have to get the T from the type's generics:

```rs
let generics = my_struct.generics().map(|g| quote { $g }).join(quote {,});
let my_type = my_struct.as_type();
quote {
    impl<$generics> Foo for $my_type { ... }
}
```

This is what `derive` does already since you don't know the names of the generics ahead of time in practice.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/comptime/value.rs [hunk 1]
```diff
@@ -497,7 +497,12 @@ impl<'value, 'interner> Display for ValuePrinter<'value, 'interner> {
             Value::Quoted(tokens) => {
                 write!(f, "quote {{")?;
                 for token in tokens.iter() {
-                    write!(f, " {token}")?;
+                    match token {
+                        Token::QuotedType(id) => {
+                            write!(f, " {:?}", self.interner.get_quoted_type(*id))?;
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/comptime/value.rs:502]:
```suggestion
                            write!(f, " {}", self.interner.get_quoted_type(*id))?;
```
^ should this be `Display`ed?

File: test_programs/execution_success/derive/src/main.nr [hunk 2]
```diff
@@ -21,15 +21,18 @@ comptime fn derive_do_nothing(s: StructDefinition) -> Quoted {
 
 // Test stdlib derive fns & multiple traits
 #[derive(Eq, Default)]
-struct MyOtherStruct {
-    field1: u32,
-    field2: u64,
+struct MyOtherStruct<A, B> {
```
Comment (michaeljklein) [inline test_programs/execution_success/derive/src/main.nr:24]: Please add a nested-struct and/or a `struct -> tuple/array/slice -> struct` example

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/hir/comptime/value.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/comptime/value.rs:502]: Yep

File: test_programs/execution_success/derive/src/main.nr (hunk 2 above)
Comment (jfecher) [reply test_programs/execution_success/derive/src/main.nr:24]: Done

**Review by michaeljklein — APPROVED** [review]

**Pull request merged**